packages/core/src/proxy_configuration.ts

```diff
@@ -84,6 +97,64 @@ export interface ProxyInfo {
     port: number | string;
 }
 
+interface TieredProxyOptions {
+    request?: Request;
+    proxyTier?: number;
+}
+
+/**
+ * Internal class for tracking the proxy tier history for a specific domain.
+ *
+ * Predicts the best proxy tier for the next request based on the error history for different proxy tiers.
+ */
+class ProxyTierTracker {
+    private histogram: number[];
+    private currentTier: number;
+
+    constructor(tieredProxyUrls: string[][]) {
+        this.histogram = tieredProxyUrls.map(() => 0);
+        this.currentTier = 0;
+    }
+
+    /**
+     * Processes a single step of the algorithm and updates the current tier prediction based on the error history.
+     */
+    private processStep(): void {
+        this.histogram.forEach((x, i) => {
+            if (this.currentTier === i) return;
+            if (x > 0) this.histogram[i]--;
+        });
+
+        const left = this.currentTier > 0 ? this.histogram[this.currentTier - 1] : Infinity;
+        const right = this.currentTier < this.histogram.length - 1 ? this.histogram[this.currentTier + 1] : Infinity;
+
+        if (this.histogram[this.currentTier] > Math.min(left, right)) {
+            this.currentTier = left <= right ? this.currentTier - 1 : this.currentTier + 1;
+        }
+
+        if (this.histogram[this.currentTier] === left) {
+            this.currentTier--;
+        }
+    }
+
+    /**
+     * Increases the error score for the given proxy tier. This raises the chance of picking a different proxy tier for the subsequent requests.
+     * @param tier The proxy tier to mark as problematic.
+     */
+    addError(tier: number) {
+        this.histogram[tier] += 10;
+    }
+
+    /**
+     * Returns the best proxy tier for the next request based on the error history for different proxy tiers.
+     * @returns The proxy tier prediction
+     */
+    getTier() {
+        this.processStep();
```

Comment:
Again, a "getter" that updates the internal state seems shady, it'd make sense to rename it to reflect that.